File: Subarray_Sum_Closest.cpp
class Solution {
public:
    /**
     * @param nums: A list of integers
     * @return: A list of integers includes the index of the first number 
     *          and the index of the last number
     */
    vector<int> subarraySumClosest(vector<int> nums){
        // write your code here
        int length = nums.size();
        vector<pair<int, int>> acc(length + 1, pair<int, int>(0, 0));
        acc[0].second = -1;//acc value, index
        for (int i = 1; i < length + 1; ++i){
            acc[i].first = acc[i - 1].first + nums[i - 1];
            acc[i].second = i - 1;
        }
        
        sort(acc.begin(), acc.end());
        
        int min_diff = INT_MAX;
        int min_index = 0;
        int max_index = 0;
        for (int i = 1; i < length + 1; ++i){
            int diff = abs(acc[i].first - acc[i - 1].first);
            if (min_diff > diff){
                min_diff = diff;
                min_index = min(acc[i].second, acc[i - 1].second) + 1;
                max_index = max(acc[i].second, acc[i - 1].second);
            }
        }
        return {min_index, max_index};
    }
};

File: Subarray Sum Closest.py
class Solution:
    """
    @param nums: A list of integers
    @return: A list of integers includes the index of the first number
             and the index of the last number
    """
    def subarraySumClosest(self, nums):
        # write your code here
        # if not nums:
        #     return []
        # length = len(nums)
        # ret = [None] * length
        # ret[0] = (nums[0], 0)
        # for i in range(1, length):
        #     ret[i] = (nums[i] + ret[i - 1][0], i)
        # # sorted(ret)
        # ret.sort()
        # distance = float('inf')
        # ans = []
        # for i in range(1, length):
        #     if distance > abs(ret[i][0] - ret[i - 1][0]):
        #         distance = ret[i][0] - ret[i - 1][0]
        #         ans = [ret[i][1], ret[i - 1][1]]
        # if ans:
        #     ans.sort()
        #     ans[0] += 1
        #     return ans
        # else:
        #     return [0, 0]

        length = len(nums)
        pre_sums = [None] * (length + 1)
        pre_sums[0] = (0, 0)
        for i, n in enumerate(nums):
            pre_sums[i + 1] = (pre_sums[i][0] + n, i + 1)

        pre_sums.sort(key=lambda x: x[0])
        ans = []
        closest = float('inf')
        for i in range(1, length + 1):
            tmp = abs(pre_sums[i][0] - pre_sums[i - 1][0])
            if closest > tmp:
                closest = tmp
                if pre_sums[i][1] > pre_sums[i - 1][1]:
                    ans = [[pre_sums[i - 1][1], pre_sums[i][1] - 1]]
                else:
                    ans = [[pre_sums[i][1], pre_sums[i - 1][1] - 1]]

            elif closest == tmp:
                if pre_sums[i][1] > pre_sums[i - 1][1]:
                    ans.append([pre_sums[i - 1][1], pre_sums[i][1] - 1])
                else:
                    ans.append([pre_sums[i][1], pre_sums[i - 1][1] - 1])

        ans.sort()
        # print(ans[-1])
        return ans[-1]
        # return ans
